File: app/scripts/process-models/automaton.js
'use strict';

const AUTOMATON = {
	get type(){
		return this.type;
	},

	get id(){
		return this.id;
	},

	get root(){
		return this.nodeMap[this.rootId];
	},

	set root(id){
		this.rootId = id;
	},

	get nodes(){
		const nodes = [];
		for(let id in this.nodeMap){
			nodes.push(this.nodeMap[id]);
		}

		return nodes;
	},

	getNode: function(id){
		return this.nodeMap[id];
	},

	addNode: function(id, metaData){
		id = (id === undefined) ? this.nextNodeId : id;
		metaData = (metaData === undefined) ? {} : metaData;
		const locationSet = {};
		locationSet[this.id] = true;
		const node = new AutomatonNode(id, {}, {}, locationSet, metaData);
		this.nodeMap[id] = node;
		this.nodeCount++;
		return node;
	},

	removeNode: function(id){
		if(this.nodeMap[id] === undefined){
			return;
		}

		delete this.nodeMap[id];

		for(let i in this.edgeMap){
			const edge = this.edgeMap[i];
			if(edge.incomingNode === id){
				this.removeEdge(i);
			}

			if(edge.outgoingNode === id){
				this.removeEdge(i);
			}
		}

		if(this.rootId === id){
			this.rootId === undefined;
		}

		this.nodeCount--;
	},

	combineNodes: function(node1, node2){
		const incoming = node2.incomingEdges;
		for(let i = 0; i < incoming.length; i++){
			const id = incoming[i];
			node1.addIncomingEdge(id);
			this.getEdge(id).to = node1.id;
		}

		const outgoing = node2.outgoingEdges;
		for(let i = 0; i < outgoing.length; i++){
			const id = outgoing[i];
			node1.addOutgoingEdge(id);
			this.getEdge(id).from = node1.id;
		}

		node1.locations = node2.locations;

		for(let key in node2.metaData){
			node1.metaData[key] = node2.metaData[key];
		}

		delete this.nodeMap[node2.id];
		this.nodeCount--;
	},

	coaccessible: function(node, label){
		const edges = node.outgoingEdges.map(id => this.getEdge(id)).filter(e => e.label === label);
		for(let i = 0; i < edges.length; i++){
			const edge = edges[i];
			const next = this.getNode(edge.to);
			edges[i] = { node:next, edge:edge };
		}

		return edges;
	},

	isCoaccessible: function(node, label){
		const edges = node.outgoingEdges.map(id => this.getEdge(id));
		for(let i = 0; i < edges.length; i++){
			if(edges[i].label === label){
				return true;
			}
		}

		return false;
	},

	get edges(){
		const edges = [];
		for(let id in this.edgeMap){
			edges.push(this.edgeMap[id]);
		}

		return edges;
	},

	getEdge: function(id){
		return this.edgeMap[id];
	},

	addEdge: function(id, label, from, to, metaData){
		const locationSet = {};
		locationSet[this.id] = true;

		if(metaData === undefined){
			metaData = {};
			const edges = this.edges.filter(e => e.label === label);
			metaData.action = edges.length;
		}

		const edge = new AutomatonEdge(id, label, from.id, to.id, locationSet, metaData);
		from.addOutgoingEdge(id);
		to.addIncomingEdge(id);
		this.edgeMap[id] = edge;
		this.edgeCount++;
		return edge;
	},

	removeEdge: function(id){
		if(this.edgeMap[id] === undefined){
			return;
		}

		delete this.edgeMap[id];
		this.edgeCount--;

		const nodes = this.nodes;
		for(let i = 0; i < nodes.length; i++){
			const node = nodes[i];
			delete node.incomingEdgeSet[id];
			delete node.outgoingEdgeSet[id];
		}
	},

	removeDuplicateEdges: function(){
		const toDelete = {};
		const edges = this.edges;
		for(let i = 0; i < edges.length; i++){
			const edge1 = edges[i];
			for(let j = i + 1; j < edges.length; j++){
				const edge2 = edges[j];
				if(toDelete[edge2.id]){
					continue;
				}

				if(edge1.from === edge2.from && edge1.to === edge2.to && edge1.label === edge2.label){
					toDelete[edge2.id] = true;
				}
			}
		}

		for(let id in toDelete){
			this.removeEdge(id);
		}
	},

	get alphabet(){
		const labels = this.edges.map(e => e.label);
		const alphabet = {};
		for(let i = 0; i < labels.length; i++){
			alphabet[labels[i]] = true;
		}


		return alphabet;
	},

	relabelEdges: function(oldLabel, newLabel){
		const edges = this.edges.filter(e => e.label === oldLabel);
		for(let i = 0; i < edges.length; i++){
			edges[i].label = newLabel;
		}
	},

	addAutomaton: function(automaton){
		const nodes = automaton.nodes;
		for(let i = 0; i < nodes.length; i++){
			const node = nodes[i];
			this.nodeMap[node.id] = node;
			this.nodeCount++;
		}

		const edges = automaton.edges;
		for(let i = 0; i < edges.length; i++){
			const edge = edges[i];
			this.edgeMap[edge.id] = edge;
			this.edgeCount++;
		}
	},

	get clone(){
		if(this.metaData.cloneCount === undefined){
			this.metaData.cloneCount = 0;
		}

		const cloneId = this.metaData.cloneCount++;

		const automaton = new Automaton(this.id + '.' + cloneId);

		// clone nodes from this automaton and add them to the clone
		const nodes = this.nodes;
		for(let i = 0; i < nodes.length; i++){
			const id = nodes[i].id + '.' + cloneId;
			const incoming = relabelSet(JSON.parse(JSON.stringify(nodes[i].incomingEdgeSet)));
			const outgoing = relabelSet(JSON.parse(JSON.stringify(nodes[i].outgoingEdgeSet)));
			const locations = JSON.parse(JSON.stringify(nodes[i].locationSet));
			const metaData = JSON.parse(JSON.stringify(nodes[i].metaData));
			const node = new AutomatonNode(id, incoming, outgoing, locations, metaData);
			automaton.nodeMap[id] = node;
			automaton.nodeCount++;

			if(nodes[i].id === this.rootId){
				automaton.root = id;
				node.metaData.startNode = true;
			}
		}

		// clone edges from this automaton and add them to the clone
		const edges = this.edges;
		for(let i = 0; i < edges.length; i++){
			const id = edges[i].id + '.' + cloneId;
			const label = edges[i].label;
			const from = edges[i].from + '.' + cloneId;
			const to = edges[i].to + '.' + cloneId;
			const locations = JSON.parse(JSON.stringify(edges[i].locationSet));
			const metaData = JSON.parse(JSON.stringify(edges[i].metaData));
			const edge = new AutomatonEdge(id, label, from, to, locations, metaData);
			automaton.edgeMap[id] = edge;
			automaton.edgeCount++;
		}

		return automaton;

		function relabelSet(set){
			const newSet = {};
			for(let label in set){
				newSet[label + '.' + cloneId] = true;
			}

			return newSet;
		}
	},

	trim: function(){
		const visited = {};
		const fringe = [this.root];
		while(fringe.length !== 0){
			const current = fringe.pop();
			if(visited[current.id]){
				continue;
			}

			visited[current.id] = true;

			const neighbours = current.outgoingEdges.map(id => this.getNode(this.getEdge(id).to));
			for(let i = 0; i < neighbours.length; i++){
				const neighbour = neighbours[i];
				if(!visited[neighbour.id]){
					fringe.push(neighbour);
				}
			}

			if(neighbours.length === 0 && current.metaData.isTerminal === undefined){
			  current.metaData.isTerminal = 'stop';
      }
		}

		for(let id in this.nodeMap){
			if(!visited[id]){
				this.removeNode(id);
			}
		}
	},

	get nextNodeId(){
		return this.id + '.n' + this.nodeId++;
	},

	get nextEdgeId(){
		return this.id + '.e' + this.edgeId++;
	},

	// Global Functions

	convert: function(automaton){
		// check that the object has the correct properties
		const properties = ['type', 'id', 'rootId', 'nodeMap', 'edgeMap', 'metaData'];
		let match = true;
		for(let i = 0; i < properties.length; i++){
			if(!automaton.hasOwnProperty(properties[i])){
				match = false;
				break;
			}
		}

		if(match){
			for(let id in automaton.nodeMap){
				const node = automaton.nodeMap[id];
				const nodeProperties = ['id', 'incomingEdgeSet', 'outgoingEdgeSet', 'locationSet', 'metaData'];
				let nodeMatch = true;
				for(let i = 0; i < nodeProperties.length; i++){
					if(!node.hasOwnProperty(nodeProperties[i])){
						nodeMatch = false;
						break;
					}
				}

				if(!nodeMatch){
					// throw error
				}

				Object.setPrototypeOf(node, AUTOMATON_NODE);
				automaton.nodeMap[id] = node;
			}

			for(let id in automaton.edgeMap){
				const edge = automaton.edgeMap[id];
				const edgeProperties = ['id', 'label', 'from', 'to', 'locationSet', 'metaData'];
				let edgeMatch = true;
				for(let i = 0; i < edgeProperties.length; i++){
					if(!edge.hasOwnProperty(edgeProperties[i])){
						edgeMatch = false;
						break;
					}
				}

				if(!edgeMatch){
					// throw error
				}

				Object.setPrototypeOf(edge, AUTOMATON_EDGE);
				automaton.edgeMap[id] = edge;
			}

			Object.setPrototypeOf(automaton, AUTOMATON);
			return automaton;
		}
	}
};

function Automaton(id){
	this.type = 'automata';
	this.id = id;
	this.nodeMap = {};
	this.nodeCount = 0;
	this.edgeMap = {};
	this.edgeCount = 0;
	this.nodeId = 0;
	this.edgeId = 0;
	this.metaData = {};
	Object.setPrototypeOf(this, AUTOMATON);
}

const AUTOMATON_NODE = {
	get type(){
		return 'node';
	},

	get id(){
		return id;
	},

	get incomingEdges(){
		const incoming = [];
		for(let id in this.incomingEdgeSet){
			incoming.push(id);
		}

		return incoming;
	},

	addIncomingEdge: function(id){
		this.incomingEdgeSet[id] = true;
	},

	removeIncomingEdge: function(id){
		delete this.incomingEdgeSet[id];
	},

	get outgoingEdges(){
		const outgoing = [];
		for(let id in this.outgoingEdgeSet){
			outgoing.push(id);
		}

		return outgoing;
	},

	addOutgoingEdge: function(id){
		this.outgoingEdgeSet[id] = true;
	},

	removeOutgoingEdge: function(id){
		delete this.outgoingEdgeSet[id];
	},

	get locations(){
		return JSON.parse(JSON.stringify(this.locationSet));
	},

	set locations(locations){
		this.locationSet = {};
		for(let id in locations){
			this.locationSet[id] = true;
		}
	},

	addLocation: function(location){
		this.locationSet[location] = true;
	},

	removeLocation: function(location){
		delete this.locationSet[location];
	},

	isTerminal: function(){
		for(let id in this.outgoingEdgeSet){
			return false;
		}

		return true;
	},

	isUnreachable: function(){
		for(let id in this.incomingEdgeSet){
			return false;
		}

		return true;
	},

	addMetaData: function(key, data){
		this.metaData[key] = data;
	},

	getMetaData: function(key){
		return this.metaData[key];
	},

	deleteMetaData: function(key){
		delete this.metaData[key];
	}
};

function AutomatonNode(id, incomingEdges, outgoingEdges, locations, metaData){
	this.id = id;
	this.incomingEdgeSet = (incomingEdges === undefined) ? {} : incomingEdges;
	this.outgoingEdgeSet = (outgoingEdges === undefined) ? {} : outgoingEdges;
	this.locationSet = (locations === undefined) ? {} : locations;
	this.metaData = (metaData === undefined) ? {} : metaData;
	Object.setPrototypeOf(this, AUTOMATON_NODE);
}

const AUTOMATON_EDGE = {
	get type(){
		return 'edge';
	},

	get id(){
		return this.id;
	},

	get label(){
		return this.label;
	},

	set label(label){
		this.label = label;
	},

	get incomingNode(){
		return this.from;
	},

	get outgoingNode(){
		return this.to;
	},

	get locations(){
		return JSON.parse(JSON.stringify(this.locationSet));
	},

	set locations(locations){
		this.locationSet = {};
		for(let id in locations){
			this.locationSet[id] = true;
		}
	},

	addLocation: function(location){
		this.locationSet[location] = true;
	},

	removeLocation: function(location){
		delete this.locationSet[location];
	},

	isHidden: function(){
		return this.label === TAU;
	},

	isDeadlocked: function(){
		return this.label === DELTA;
	},

	addMetaData: function(key, data){
		this.metaData[key] = data;
	},

	getMetaData: function(key){
		return this.metaData[key];
	},

	deleteMetaData: function(key){
		delete this.metaData[key];
	}
};

function AutomatonEdge(id, label, from, to, locations, metaData){
	this.id = id;
	this.label = label;
	this.from = from;
	this.to = to;
	this.locationSet = (locations === undefined) ? {} : locations;
	this.metaData = (metaData === undefined) ? {} : metaData;
	Object.setPrototypeOf(this, AUTOMATON_EDGE);
}

File: modelchecker/src/main/java/mc/process_models/automata/Automaton.java
package mc.process_models.automata;

import com.google.gson.Gson;
import guru.nidi.graphviz.attribute.Rank;
import guru.nidi.graphviz.attribute.RankDir;
import guru.nidi.graphviz.engine.*;
import guru.nidi.graphviz.model.Label;
import guru.nidi.graphviz.model.Link;
import guru.nidi.graphviz.model.MutableGraph;
import guru.nidi.graphviz.model.Node;
import mc.compiler.Guard;
import mc.exceptions.CompilationException;
import mc.process_models.ProcessModel;
import mc.process_models.ProcessModelObject;
import mc.util.GraphvizV8ThreadedEngine;
import mc.util.Location;
import mc.util.expr.ExpressionSimplifier;
import mc.util.expr.OrOperator;
import mc.webserver.webobjects.LogMessage;
import org.json.JSONArray;
import org.json.JSONObject;

import java.awt.geom.Point2D;
import java.io.File;
import java.io.IOException;
import java.lang.reflect.Field;
import java.util.*;
import java.util.concurrent.BlockingQueue;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static guru.nidi.graphviz.model.Factory.mutGraph;
import static guru.nidi.graphviz.model.Factory.node;

public class Automaton extends ProcessModelObject implements ProcessModel {

    public static final boolean CONSTRUCT_ROOT = true;

    private AutomatonNode root;
    private Map<String, AutomatonNode> nodeMap;
    private Map<String, AutomatonEdge> edgeMap;
    private Map<String, List<AutomatonEdge>> alphabet;

    private int nodeId;
    private int edgeId;

    public Automaton(String id) {
        super(id, "automata");
        setupAutomaton();

        // setup the root for this automaton
        this.root = addNode();
        root.addMetaData("startNode", true);
    }

    public Automaton(String id, boolean constructRoot) {
        super(id, "automata");
        setupAutomaton();

        // only construct a root node if specified to do so
        if (constructRoot) {
            this.root = addNode();
            root.addMetaData("startNode", true);
        }
    }
    private void setupAutomaton() {
        this.nodeMap = new HashMap<>();
        this.edgeMap = new HashMap<>();
        this.alphabet = new HashMap<>();

        this.nodeId = 0;
        this.edgeId = 0;
    }

    public AutomatonNode getRoot() {
        return root;
    }
    public void position(BlockingQueue<Object> messageQueue) throws CompilationException {
        messageQueue.add(new LogMessage("Performing layout algorithm: "+getId(),true,false));
        MutableGraph g = mutGraph(getId()).setDirected().generalAttrs().add(RankDir.LEFT_TO_RIGHT);
        Map<String,Node> graphNodes = new HashMap<>();
        for (AutomatonNode node : getNodes()) {
            Node n = node(node.getId());
            if (node.hasMetaData("startNode")) {
                n = n.with(Rank.MIN);
            }
//            for (String key: node.getMetaDataKeys()) {
//                n = n.with(key,node.getMetaData(key));
//            }
            graphNodes.put(node.getId(),n);
            g.add(n);
        }
        for (AutomatonEdge edge : getEdges()) {
            Node to = graphNodes.get(edge.getTo().getId());
            Node from = graphNodes.get(edge.getFrom().getId());
            Node l = from.link(to.linkTo().with(Label.of(edge.getLabel())));
//            for (String key: edge.getMetaDataKeys()) {
//                l = l.with(key,edge.getMetaData(key));
//            }
            g.add(l);
        }
        if (Thread.currentThread().isInterrupted()) {
            throw new RuntimeException(new InterruptedException("Interrupted!"));
        }
        Graphviz viz = Graphviz.fromGraph(g);
        JSONObject obj = new JSONObject(viz.engine(Engine.DOT).render(Format.JSON).toString());
        JSONArray objects = obj.getJSONArray("objects");
        for (int i = 0; i < objects.length(); i++) {
            JSONObject ele = objects.getJSONObject(i);
            String[] posString = ele.getString("pos").split(",");

            getNode(ele.getString("name")).addMetaData("pos",new Point2D.Double(Double.parseDouble(posString[0]),Double.parseDouble(posString[1])));
        }
    }
    public void setRoot(AutomatonNode root) throws CompilationException {
        // check the the new root is defined
        if (root == null) {
            throw new CompilationException(getClass(),"Unable to set the root node to null",(Location)getMetaData().get("location"));
        }

        // check that the new root is part of this automaton
        if (!nodeMap.containsKey(root.getId())) {
            throw new CompilationException(getClass(),"Unable to set the root node to "+root.getId()+", as the root is not a part of this automaton",(Location)getMetaData().get("location"));
        }

        this.root = root;
    }

    public String getRootId() {
        return root.getId();
    }

    public List<AutomatonNode> getNodes() {
        return nodeMap.entrySet().stream()
            .map(Map.Entry::getValue)
            .collect(Collectors.toList());
    }

    public AutomatonNode getNode(String id) throws CompilationException {
        if (nodeMap.containsKey(id)) {
            return nodeMap.get(id);
        }

        throw new CompilationException(getClass(),"Unable to get the node "+id+" as it does not exist.",(Location)getMetaData().get("location"));
    }

    public AutomatonNode addNode() {
        String id = getNextNodeId();
        while(nodeMap.containsKey(id)){
            id = getNextNodeId();
        }

        return addNode(id);
    }

    public AutomatonNode addNode(String id) {
        AutomatonNode node = new AutomatonNode(id);
        nodeMap.put(id, node);
        return node;
    }

    public boolean removeNode(AutomatonNode node) {
        // check that the specified node is part of this automaton
        if (!nodeMap.containsKey(node.getId())) {
            return false;
        }
        // remove the edges that reference the specified node
        List<AutomatonEdge> edges = node.getIncomingEdges();
        edges.addAll(node.getOutgoingEdges());
        edges.stream()
            .map(ProcessModelObject::getId)
            .forEach(id -> edgeMap.remove(id));
        nodeMap.remove(node.getId());
        return true;
    }

    public AutomatonNode combineNodes(AutomatonNode node1, AutomatonNode node2) throws CompilationException {
        if(!nodeMap.containsKey(node1.getId())){
            throw new CompilationException(getClass(), node1.getId() + " was not found in the automaton " + getId(), (Location)getMetaData("location"));
        }
        if(!nodeMap.containsKey(node2.getId())){
            throw new CompilationException(getClass(),node2.getId() + " was not found in the automaton "+ getId(), (Location)getMetaData("location"));
        }
        AutomatonNode node = addNode();

        for (AutomatonEdge edge1 : node1.getIncomingEdges()) {
            for (AutomatonEdge edge2: node2.getIncomingEdges()) {
                processGuards(edge1,edge2);
            }
        }

        for (AutomatonEdge edge1 : node1.getOutgoingEdges()) {
            for (AutomatonEdge edge2: node2.getOutgoingEdges()) {
                processGuards(edge1,edge2);
            }
        }
        // add the incoming and outgoing edges from both nodes to the combined nodes
        processIncomingEdges(node, node1);
        processIncomingEdges(node, node2);
        processOutgoingEdges(node, node1);
        processOutgoingEdges(node, node2);
        // create a union of the metadata from both nodes
        for(String key : node1.getMetaDataKeys()){
            node.addMetaData(key, node1.getMetaData(key));
        }
        for(String key : node2.getMetaDataKeys()){
            node.addMetaData(key, node2.getMetaData(key));
        }
        node.removeMetaData("variables");
        Map<?,?> vars1 = (Map) node1.getMetaData("variables");
        Map<?,?> vars2 = (Map) node2.getMetaData("variables");
        if (Objects.equals(vars1,vars2)) {
            node.addMetaData("variables", vars1);
        }

        if(node1.hasMetaData("startNode") || node2.hasMetaData("startNode")){
            setRoot(node);
            node.addMetaData("startNode",true);
        }
        removeNode(node1);
        removeNode(node2);
        return node;
    }
    private void processGuards(AutomatonEdge edge1, AutomatonEdge edge2) throws CompilationException {
        if (edge1.getLabel().equals(edge2.getLabel()) && edge1.hasMetaData("guard") && edge2.hasMetaData("guard")) {
            Guard guard1 = (Guard) edge1.getMetaData("guard");
            Guard guard2 = (Guard) edge2.getMetaData("guard");
            if (guard1 == null || guard2 == null || guard1.getGuard() == null || guard2.getGuard() == null) return;
            //Since assignment should be the same (same colour) we can just copy most data from either guard.
            Guard combined = guard1.copy();
            //By putting both equations equal to eachother, if we have multiple or operations, then if one matches then it will be solveable.
            if (!guard1.getVariables().isEmpty() && !ExpressionSimplifier.equate(guard1,guard2))
                //We could take either path
                combined.setGuard(new OrOperator(guard1.getGuard(), guard2.getGuard()));
            else
                combined.setGuard(guard1.getGuard());
            edge1.addMetaData("guard",combined);
            edge2.addMetaData("guard",combined);
        }
    }
    private void processIncomingEdges(AutomatonNode node, AutomatonNode oldNode) {
        List<AutomatonEdge> edges = oldNode.getIncomingEdges();
        for (AutomatonEdge edge : edges) {
            node.addIncomingEdge(edge);
            edge.setTo(node);
            oldNode.removeIncomingEdge(edge);
        }
    }

    private void processOutgoingEdges(AutomatonNode node, AutomatonNode oldNode) {
        List<AutomatonEdge> edges = oldNode.getOutgoingEdges();
        for (AutomatonEdge edge : edges) {
            node.addOutgoingEdge(edge);
            edge.setFrom(node);
            oldNode.removeOutgoingEdge(edge);
        }
    }

    public int getNodeCount() {
        return nodeMap.size();
    }

    public List<AutomatonEdge> getEdges() {
        return edgeMap.entrySet().stream()
            .map(Map.Entry::getValue)
            .collect(Collectors.toList());
    }

    public AutomatonEdge getEdge(String id) throws CompilationException {
        if (edgeMap.containsKey(id)) {
            return edgeMap.get(id);
        }

        throw new CompilationException(getClass(),"Edge "+id+" was not found in the automaton "+getId(),(Location)getMetaData().get("location"));
    }

    public AutomatonEdge addEdge(String label, AutomatonNode from, AutomatonNode to, Map<String,Object> metaData) throws CompilationException {
        String id = getNextEdgeId();
        return addEdge(id, label, from, to, metaData);
    }

    public AutomatonEdge addEdge(String id, String label, AutomatonNode from, AutomatonNode to, Map<String,Object> metaData) throws CompilationException {
        // check that the nodes have been defined
        if (from == null) {
            throw new CompilationException(getClass(),"Unable to add the specified edge as the source was null.",(Location)getMetaData().get("location"));
        }

        if (to == null) {
            throw new CompilationException(getClass(),"Unable to add the specified edge as the destination was null.",(Location)getMetaData().get("location"));
        }

        // check that the nodes are part of this automaton
        if (!nodeMap.containsKey(from.getId())) {
            throw new CompilationException(getClass(),"Unable to add the specified edge as "+from.getId()+" is not a part of this automaton. \nPlease make sure you aren't linking directly to a parallel composed process!",(Location)getMetaData().get("location"));
        }

        if (!nodeMap.containsKey(to.getId())) {
            throw new CompilationException(getClass(),"Unable to add the specified edge as "+to.getId()+" is not a part of this automaton.  \nPlease make sure you aren't linking directly to a parallel composed process!",(Location)getMetaData().get("location"));
        }

        // check if there is already an identical edge between the specified nodes
        List<AutomatonEdge> edges = from.getOutgoingEdges().stream()
            .filter(edge -> edge.getLabel().equals(label) && edge.getTo().getId().equals(to.getId()))
            .collect(Collectors.toList());

        if(edges.size() > 0){
            for (AutomatonEdge edge : edges) {
                if (edge.getMetaData("guard") == null && !metaData.containsKey("guard")) {
                    return edge;
                }
                Guard guard = (Guard) edge.getMetaData("guard");
                Guard guard2 = (Guard) metaData.get("guard");
                if(guard != null && guard2 != null) {
                    if (Objects.equals(guard,guard2)) return edge;
                }
            }
        }

        AutomatonEdge edge = new AutomatonEdge(id, label, from, to);
        edge.getMetaData().putAll(metaData);

        // add edge reference to the incoming and outgoing nodes
        from.addOutgoingEdge(edge);
        to.addIncomingEdge(edge);

        // add edge to the edge and alphabet maps
        if (!alphabet.containsKey(label)) {
            alphabet.put(label, new ArrayList<>());
        }
        alphabet.get(label).add(edge);
        edgeMap.put(id, edge);

        return edge;
    }

    public boolean removeEdge(AutomatonEdge edge) {
        // check that the specified edge is part of this automaton
        if (!edgeMap.containsKey(edge.getId())) {
            return false;
        }

        // remove references to this edge
        edge.getFrom().removeOutgoingEdge(edge);
        edge.getTo().removeIncomingEdge(edge);

        edgeMap.remove(edge.getId());
        return true;
    }

    private boolean findEdgeWith(Link link, String id) {
        for (Map.Entry<String,Object> entry : link.attrs()) {
            if (entry.getKey().equals("id") && entry.getValue().equals(id)) {
                return true;
            }
        }
        return false;
    }

    public boolean removeEdge(String id) {
        if (!edgeMap.containsKey(id)) {
            return false;
        }

        return removeEdge(edgeMap.get(id));
    }

    public void relabelEdges(String oldLabel, String newLabel) {
        if (alphabet.containsKey(oldLabel)) {
            List<AutomatonEdge> edges = alphabet.get(oldLabel);
            edges.forEach(edge -> edge.setLabel(newLabel));
            if (alphabet.containsKey(newLabel)) {
                edges.addAll(alphabet.get(newLabel));
            }
            alphabet.put(newLabel, edges);
            alphabet.remove(oldLabel);
        }
    }

    public int getEdgeCount() {
        return edgeMap.size();
    }

    public Set<String> getAlphabet() {
        return alphabet.keySet();
    }

    public AutomatonNode addAutomaton(Automaton automaton) throws CompilationException {
        AutomatonNode root = null;
        for(AutomatonNode node : automaton.getNodes()){
            AutomatonNode newNode = addNode(node.getId());
            for(String key : node.getMetaDataKeys()){
                if(key.equals("startNode")){
                    root = newNode;
                    if(this.root != null){
                        continue;
                    }

                    this.root = newNode;
                }

                newNode.addMetaData(key, node.getMetaData(key));
            }
        }

        for (AutomatonEdge edge : automaton.getEdges()) {
            AutomatonNode from = getNode(edge.getFrom().getId());
            AutomatonNode to = getNode(edge.getTo().getId());
            AutomatonEdge newEdge = addEdge(edge.getId(), edge.getLabel(), from, to, edge.getMetaData());
            for (String key : edge.getMetaDataKeys()) {
                newEdge.addMetaData(key, edge.getMetaData(key));
            }
        }
        if (root == null) {
            throw new CompilationException(getClass(),"There was no root found while trying to add an automaton");
        }
        return root;
    }


    public String getNextNodeId() {
        return getId() + ".n" + nodeId++;
    }

    public String getNextEdgeId() {
        return getId() + ".e" + edgeId++;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("automaton:{\n");
        builder.append("\tnodes:{\n");
        for (AutomatonNode node : nodeMap.values()) {
            builder.append("\t\t" + node.getId());
            if (node == root) {
                builder.append("(root)");
            }
            builder.append("\n");
        }
        builder.append("\t}\n\tedges:{\n");
        for (AutomatonEdge edge : edgeMap.values()) {
            builder.append("\t\t" + edge.getFrom().getId() + " -" + edge.getLabel() + "> " + edge.getTo().getId() + "\n");
        }
        builder.append("\t}\n}");

        return builder.toString();
    }

    public Automaton copy() throws CompilationException {
        Automaton copy = new Automaton(getId(), !CONSTRUCT_ROOT);
        copy.nodeId = nodeId;
        copy.edgeId = edgeId;
        List<AutomatonNode> nodes = getNodes();
        for(AutomatonNode node : nodes){
            AutomatonNode newNode = copy.addNode(node.getId());
            for(String key : node.getMetaDataKeys()){
                newNode.addMetaData(key, node.getMetaData(key));
                if(key.equals("startNode")){
                    copy.setRoot(newNode);
                }
            }
        }

        List<AutomatonEdge> edges = getEdges();
        for(AutomatonEdge edge : edges){
            AutomatonNode from = copy.getNode(edge.getFrom().getId());
            AutomatonNode to = copy.getNode(edge.getTo().getId());
            copy.addEdge(edge.getId(), edge.getLabel(), from, to, edge.getMetaData());
        }

        for(String key : getMetaDataKeys()){
            copy.addMetaData(key, getMetaData(key));
        }

        return copy;
    }
}
